### PR TITLE
Torchserve v2 protocol

### LINKED ISSUE #1253: Support for KFServing V2 protocol for torchserve
/kind feature

Add KFServing V2 predict api protocol support for torchserve predictor.
[Predict protocol V2](https://github.com/kubeflow/kfserving/tree/master/docs/predict-api/v2)

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by jagadeeshi2i**
- Adds v2 API for Torchserve
- Adds gRPC support for Torchserve
- Deprecate KServe's pytorchserver

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1253 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

- Bump Torchserve to 0.5.0
- KServe v2 Protocol REST API support for Torchserve Inferenceservice
- Deprecate kserve's pytorchserver and support KServe's v2 REST inference protocol

```

### CODE REVIEW

File: config/configmap/inferenceservice.yaml [hunk 1]
```diff
@@ -61,24 +61,13 @@ data:
...
-          "v2" : {
-            "image": "pytorch/torchserve-kfs",
-            "defaultImageVersion": "0.4.1",
-            "defaultGpuImageVersion": "0.4.1-gpu",
-            "supportedFrameworks": [
-              "pytorch"
-            ],
-            "multiModelServer": false
-          }
+          "image": "jagadeeshj/torchserve-kfsv2",
```
Comment (pugangxa) [inline config/configmap/inferenceservice.yaml]: Better push image to the kserve repo

Comment (yuzisun) [reply config/configmap/inferenceservice.yaml]: @jagadeeshi2i When is the official image going to be released from torchserve?

File: docs/samples/README.md [hunk 2]
```diff
@@ -26,14 +26,14 @@ After models are deployed onto model servers with KFServing, you get all the fol
...
 | [Triton Inference Server](https://github.com/triton-inference-server/server) | [TensorFlow,TorchScript,ONNX,TensorRT](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/model_repository.html)| v2 | :heavy_check_mark: | :heavy_check_mark: | [Compatibility Matrix](https://docs.nvidia.com/deeplearning/frameworks/support-matrix/index.html)| [Triton Examples](./v1beta1/triton) |
 | [TFServing](https://www.tensorflow.org/tfx/guide/serving) | [TensorFlow SavedModel](https://www.tensorflow.org/guide/saved_model) | v1 | :heavy_check_mark: | :heavy_check_mark: | [TFServing Versions](https://github.com/tensorflow/serving/releases) | [TensorFlow Examples](./v1alpha2/tensorflow)  |
-| [TorchServe](https://pytorch.org/serve/server.html) | [Eager Model/TorchScript](https://pytorch.org/docs/master/generated/torch.save.html) | v1 | :heavy_check_mark: | :heavy_check_mark: | 0.4.1 | [TorchServe Examples](./v1beta1/torchserve)  |
+| [TorchServe](https://pytorch.org/serve/server.html) | [Eager Model/TorchScript](https://pytorch.org/docs/master/generated/torch.save.html) | v2 | :heavy_check_mark: | :heavy_check_mark: | 0.4.1 | [TorchServe Examples](./v1beta1/torchserve)  |
```
Comment (yuzisun) [inline docs/samples/README.md]:
```suggestion
| [TorchServe](https://pytorch.org/serve/server.html) | [Eager Model/TorchScript](https://pytorch.org/docs/master/generated/torch.save.html) | v1/v2 | :heavy_check_mark: | :heavy_check_mark: | 0.4.1 | [TorchServe Examples](./v1beta1/torchserve)  |
```

File: docs/samples/v1beta1/torchserve/v2/README.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
+# Predict on a InferenceService using TorchServe
+
+In this example, we use a trained pytorch mnist model to predict handwritten digits by running an inference service with [TorchServe](https://github.com/pytorch/serve) predictor.
+
+## Setup
+
+1. Your ~/.kube/config should point to a cluster with [KFServing installed](https://github.com/kubeflow/kfserving/#install-kfserving).
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/v2/README.md]:
```suggestion
1. Your ~/.kube/config should point to a cluster with [KServe installed](https://github.com/kserve/kserve/#installation).
```

File: docs/samples/v1beta1/torchserve/v2/README.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+For deploying the `InferenceService` on CPU
+```bash
+kubectl apply -f torchserve-v2.yaml
+```
+
+Expected Output
+
+```bash
+$inferenceservice.serving.kubeflow.org/torchserve-v2 created
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/v2/README.md]:
```suggestion
$inferenceservice.serving.kserve.io/torchserve-v2 created
```

File: test/e2e/transformer/test_transformer.py [hunk 5]
```diff
@@ -36,21 +35,22 @@ def test_transformer():
...
         min_replicas=1,
         pytorch=V1beta1TorchServeSpec(
-            storage_uri='gs://kfserving-examples/models/torchserve/image_classifier',
+            storage_uri="gs://kfserving-examples/models/torchserve/image_classifier",
+            protocol_version="v1",
             resources=V1ResourceRequirements(
-                requests={'cpu': '100m', 'memory': '1Gi'},
-                limits={'cpu': '100m', 'memory': '1Gi'}
-            )
+                requests={"cpu": "1", "memory": "4Gi"},
```
Comment (yuzisun) [inline test/e2e/transformer/test_transformer.py]: Do we need to bump up the resource?

File: docs/samples/README.md [hunk 6]
```diff
@@ -25,8 +25,8 @@ After models are deployed onto model servers with KServe, you get all the follow
...
-| [TFServing](https://www.tensorflow.org/tfx/guide/serving) | [TensorFlow SavedModel](https://www.tensorflow.org/guide/saved_model) | v1 | :heavy_check_mark: | :heavy_check_mark: | [TFServing Versions](https://github.com/tensorflow/serving/releases) | [TensorFlow Examples](./v1beta1/tensorflow)  |
-| [TorchServe](https://pytorch.org/serve/server.html) | [Eager Model/TorchScript](https://pytorch.org/docs/master/generated/torch.save.html) | v1 | :heavy_check_mark: | :heavy_check_mark: | 0.4.1 | [TorchServe Examples](./v1beta1/torchserve)  |
+| [TFServing](https://www.tensorflow.org/tfx/guide/serving) | [TensorFlow SavedModel](https://www.tensorflow.org/guide/saved_model) | v1 | :heavy_check_mark: | :heavy_check_mark: | [TFServing Versions](https://github.com/tensorflow/serving/releases) | [TensorFlow Examples](./v1alpha2/tensorflow)  |
```
Comment (yuzisun) [inline docs/samples/README.md]:
```suggestion
| [TFServing](https://www.tensorflow.org/tfx/guide/serving) | [TensorFlow SavedModel](https://www.tensorflow.org/guide/saved_model) | v1 | :heavy_check_mark: | :heavy_check_mark: | [TFServing Versions](https://github.com/tensorflow/serving/releases) | [TensorFlow Examples](./v1beta1/tensorflow)  |
```

File: config/configmap/inferenceservice.yaml (hunk 1 above)
Comment (jagadeeshi2i) [reply config/configmap/inferenceservice.yaml]: @yuzisun as discussed will move to kserve repo temporarily

File: test/e2e/transformer/test_transformer.py (hunk 5 above)
Comment (jagadeeshi2i) [reply test/e2e/transformer/test_transformer.py]: Made this change for dev build. Will revert them.

File: docs/samples/v1beta1/torchserve/v2/bert/bert.yaml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+apiVersion: serving.kubeflow.org/v1beta1
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/v2/bert/bert.yaml]:
```suggestion
apiVersion: serving.kserve.io/v1beta1
```

File: docs/samples/v1beta1/torchserve/v2/tensor_conv/README.md [hunk 8]
```diff
@@ -0,0 +1,17 @@
+# Generate tensor input for kfserving v2 protocol
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/v2/tensor_conv/README.md]:
```suggestion
# Generate tensor input for KServe v2 protocol
```

File: test/e2e/transformer/test_transformer.py (hunk 5 above)
Comment (yuzisun) [reply test/e2e/transformer/test_transformer.py]: Can we revert the resource requirement unless this is required?

File: test/e2e/predictor/test_torchserve.py [hunk 9]
```diff
@@ -58,3 +58,34 @@ def test_torchserve_kserve():
     res = predict(service_name, "./data/torchserve_input.json")
     assert(res.get("predictions")[0] == 2)
     kserve_client.delete(service_name, KSERVE_TEST_NAMESPACE)
+
+
+def test_torchserve_v2_kserve():
```
Comment (yuzisun) [inline test/e2e/predictor/test_torchserve.py:70]: Can we add the gRPC test as well?

Comment (jagadeeshi2i) [reply test/e2e/predictor/test_torchserve.py:70]: This requires proto files. Shall i upload it to data folder along with json input files.

Comment (yuzisun) [reply test/e2e/predictor/test_torchserve.py:70]: alternatively you can write a gRPC python client

File: docs/samples/v1beta1/torchserve/README.md [hunk 10]
```diff
@@ -45,131 +45,62 @@ via local socket.
 
 
 ## TorchServe with KFS envelope inference endpoints
-The KFServing/TorchServe integration supports KFServing v1 protocol and we are working on to support v2 protocol.
+The KFServing/TorchServe integration supports KFServing v1 and v2 protocols.
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/README.md]:
```suggestion
The KServe/TorchServe integration supports both v1 and v2 protocols.
```

File: docs/samples/v1beta1/torchserve/README.md [hunk 11]
```diff
@@ -45,131 +45,62 @@ via local socket.
...
-For deploying the `InferenceService` on GPU
-```bash
-kubectl apply -f gpu.yaml
-```
+[Sample requests for text and image classification](https://github.com/pytorch/serve/tree/master/kubernetes/kfserving/kf_request_json)
 
-Expected Output
+## Autoscaling
+One of the main serverless inference features is to automatically scale the replicas of an `InferenceService` matching the incoming workload.
+KFServing by default enables [Knative Pod Autoscaler](https://knative.dev/docs/serving/autoscaling/) which watches traffic flow and scales up and down
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/README.md]:
```suggestion
KServe by default enables [Knative Pod Autoscaler](https://knative.dev/docs/serving/autoscaling/) which watches traffic flow and scales up and down
```

File: docs/samples/v1beta1/torchserve/README.md [hunk 12]
```diff
@@ -45,131 +45,62 @@ via local socket.
...
-```bash
-MODEL_NAME=mnist
-SERVICE_HOSTNAME=$(kubectl get inferenceservice torchserve -o jsonpath='{.status.url}' | cut -d "/" -f 3)
-```
+## Monitoring
+[Expose metrics and setup grafana dashboards](metrics/README.md)
 
-Use [image converter](../imgconv/README.md) to create input request for mnist. 
-For other models please refer to [input request](https://github.com/pytorch/serve/tree/master/kubernetes/kfserving/kf_request_json)
+## gRPC
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/README.md:82]: Do we support http based v2 protocol or only gRPC?

File: docs/samples/v1beta1/torchserve/grpc.yaml [hunk 13]
```diff
@@ -0,0 +1,12 @@
+apiVersion: serving.kubeflow.org/v1beta1
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/grpc.yaml]:
```suggestion
apiVersion: serving.kserve.io/v1beta1
```

File: docs/samples/v1beta1/torchserve/v1/README.md [hunk 14]
```diff
@@ -0,0 +1,129 @@
+# Predict on a InferenceService using TorchServe
+
+In this example, we use a trained pytorch mnist model to predict handwritten digits by running an inference service with [TorchServe](https://github.com/pytorch/serve) predictor.
+
+## Setup
+
+1. Your ~/.kube/config should point to a cluster with [KFServing installed](https://github.com/kubeflow/kfserving/#install-kfserving).
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/v1/README.md]:
```suggestion
1. Your ~/.kube/config should point to a cluster with [KServe installed](https://github.com/kserve/kserve#installation).
```

File: docs/samples/v1beta1/torchserve/v1/README.md [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+
+For deploying the `InferenceService` on GPU
+```bash
+kubectl apply -f gpu.yaml
+```
+
+Expected Output
+
+```bash
+$inferenceservice.serving.kubeflow.org/torchserve created
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/v1/README.md]:
```suggestion
$inferenceservice.serving.kserve.io/torchserve created
```

File: install/v0.5.0-rc2/kfserving.yaml [hunk 16]
```diff
@@ -1358,8 +1358,6 @@ spec:
                         type: object
                       pytorch:
                         properties:
-                          modelClassName:
```
Comment (yuzisun) [inline install/v0.5.0-rc2/kfserving.yaml:1361]: Do not need to modify the old release artifacts

File: test/e2e/predictor/test_torchserve.py [hunk 17]
```diff
@@ -37,24 +39,108 @@ def test_torchserve_kserve():
...
+
+    kserve_client.create(isvc)
+    kserve_client.wait_isvc_ready(service_name, namespace=KSERVE_TEST_NAMESPACE)
+
+    with open("./data/torchserve_input.json", 'rb') as f:
+        data = f.read()
+
+    input_data = {'data': data}
+    stub = grpc_stub(service_name, KSERVE_TEST_NAMESPACE)
+    response = stub.Predictions(
```
Comment (yuzisun) [inline test/e2e/predictor/test_torchserve.py:141]: This looks like torchserve gRPC API, not kserve's v2 gRPC ?

File: docs/samples/v1beta1/torchserve/README.md [hunk 18]
```diff
@@ -45,131 +45,62 @@ via local socket.
...
-## Canary Rollout
-Canary rollout is a deployment strategy when you release a new version of model to a small percent of the production traffic.
+#### Run gRPC Inference
 
-[Canary Deployment](canary/README.md)
-
-## Monitoring
-[Expose metrics and setup grafana dashboards](metrics/README.md)
+```bash
+python torchserve_gprc_client.py infer <model_name> <input json>
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/README.md]:
```suggestion
python torchserve_grpc_client.py infer <model_name> <input json>
```

File: test/e2e/predictor/test_torchserve.py [hunk 19]
```diff
@@ -37,24 +39,108 @@ def test_torchserve_kserve():
...
+    kserve_client.delete(service_name, KSERVE_TEST_NAMESPACE)
+
+
+def test_torchserve_grpc():
+    service_name = "mnist"
+    predictor = V1beta1PredictorSpec(
+        min_replicas=1,
+        pytorch=V1beta1TorchServeSpec(
+            storage_uri="gs://kfserving-examples/models/torchserve/image_classifier/",
+            ports=[V1ContainerPort(container_port=7070, name="h2c", protocol="TCP")],
```
Comment (yuzisun) [inline test/e2e/predictor/test_torchserve.py:112]: where is 7070 configured ? I do not find it anywhere in the configuration file

File: docs/samples/v1beta1/torchserve/v1/config.properties [hunk 20]
```diff
@@ -1,10 +1,13 @@
 inference_address=http://0.0.0.0:8085
-management_address=http://0.0.0.0:8081
+management_address=http://0.0.0.0:8085
```
Comment (yuzisun) [inline docs/samples/v1beta1/torchserve/v1/config.properties:2]: is `management_address` supposed to be same as the `inference_address` now?

File: test/e2e/transformer/test_transformer.py [hunk 21]
```diff
@@ -36,21 +35,22 @@ def test_transformer():
...
+                limits={"cpu": "1", "memory": "4Gi"},
+            ),
         ),
     )
     transformer = V1beta1TransformerSpec(
         min_replicas=1,
         containers=[V1Container(
-                      image='809251082950.dkr.ecr.us-west-2.amazonaws.com/kserve/image-transformer:latest',
-                      name='kserve-container',
+                      image='kserve/torchserve-image-transformer:latest',
```
Comment (yuzisun) [inline test/e2e/transformer/test_transformer.py]: any reason to change the test image?

File: test/e2e/transformer/test_transformer.py [hunk 22]
```diff
@@ -36,21 +35,22 @@ def test_transformer():
...
+                      name='kserve-transformer',
                       resources=V1ResourceRequirements(
-                          requests={'cpu': '100m', 'memory': '1Gi'},
-                          limits={'cpu': '100m', 'memory': '1Gi'}),
+                          requests={'cpu': '1', 'memory': '4Gi'},
```
Comment (yuzisun) [inline test/e2e/transformer/test_transformer.py]: change back the resource requirement?

File: test/e2e/predictor/test_torchserve.py (hunk 19 above)
Comment (jagadeeshi2i) [reply test/e2e/predictor/test_torchserve.py:112]: It is the default port for Torchserve gRPC

File: docs/samples/v1beta1/torchserve/v1/config.properties (hunk 20 above)
Comment (jagadeeshi2i) [reply docs/samples/v1beta1/torchserve/v1/config.properties:2]: With latest Torchserve release both management and inference APIs can be accessed via same port.

File: test/e2e/predictor/test_torchserve.py (hunk 17 above)
Comment (jagadeeshi2i) [reply test/e2e/predictor/test_torchserve.py:141]: Yes. Adding support for KServe gRPC in a separate PR.

**Pull request merged**